### PR TITLE
implement cache

### PR DESCRIPTION
**Pull request by mikeclimbrock**

### CODE REVIEW

**Comment by paololim** [comment]
Awesome!

lgtm - ok, i think this will work now

File: app/io/flow/event/Queue.scala [hunk 1]
```diff
@@ -221,32 +227,43 @@ case class KinesisStream(
...
-        .withShardIteratorType(ShardIteratorType.TRIM_HORIZON)
-
-      val shardIterator = withErrorHandler("getShardIterator") {
-        kinesisClient.getShardIterator(request).getShardIterator
-      }
+    if (!shardIteratorMap.contains(shardId)) {
+      getShardIterator(shardId, ShardIteratorType.TRIM_HORIZON)
 
-      shardSequenceNumberMap += (shardId -> shardIterator)
+    } else if (shardIteratorMap(shardId).timestamp.isBefore(DateTime.now().minusMinutes(3))) {
```
Comment (mbryzek) [inline app/io/flow/event/Queue.scala]: why 3 minutes?

Comment (mikeclimbrock) [inline app/io/flow/event/Queue.scala]: actually bumped it to 4 (in master).

This is based off the fact that Kinesis expires the `shardIterator` after 5 minutes

Documentation states: "A shard iterator expires five minutes after it is returned to the requester."

Comment (mbryzek) [inline app/io/flow/event/Queue.scala]: maybe clearer for readers to ```
val ShardExpirationTimeMinutes = 5

} else if (shardIteratorMap(shardId).timestamp.isBefore(DateTime.now().minusMinutes(ShardExpirationTimeMinutes-1))) {
```

or something to document the fact

Comment (mikeclimbrock) [inline app/io/flow/event/Queue.scala]: we could, worth noting I did put it in the method comments:
https://github.com/flowcommerce/lib-event/blob/master/app/io/flow/event/Queue.scala#L231

**Pull request merged**